src/audio/kpb.c

```diff
@@ -749,12 +752,69 @@ static void kpb_init_draining(struct comp_data *kpb, struct kpb_client *cli)
 	}
 }
 
+/**
+ * \brief Draining task.
+ *
+ * \param[in] arg - pointer keeping drainig data previously prepared
+ * by kpb_init_draining().
+ *
+ * \return none.
+ */
 static uint64_t kpb_draining_task(void *arg)
 {
-	/* TODO: while loop drainning history buffer accoriding to
-	 * clients request
-	 */
-	return 0;
+	struct dd *draining_data = (struct dd *)arg;
+	struct comp_buffer *sink = draining_data->sink;
+	struct hb *buff = draining_data->history_buffer;
+	size_t history_depth = draining_data->history_depth;
+	size_t size_to_read;
+	size_t size_to_copy;
+	uint32_t draining_done = 0;
+	uint32_t move_buffer = 0;
+	uint64_t ret = 1;
+
+	trace_kpb("kpb_draining_task(), start.");
+
+	while (history_depth > 0 && !draining_done) {
+		size_to_read = (uint32_t)buff->end_addr - (uint32_t)buff->r_ptr;
+
+		if (size_to_read > sink->free) {
+			if (sink->free >= history_depth) {
+				size_to_copy = history_depth;
+				draining_done = 1;
+			} else {
+				size_to_copy = sink->free;
+			}
+		} else {
+			if (size_to_read >= history_depth) {
+				size_to_copy = history_depth;
+				draining_done = 1;
+			} else {
+				size_to_copy = size_to_read;
+				move_buffer = 1;
+			}
+		}
+
+		memcpy(sink->w_ptr, buff->r_ptr, size_to_copy);
+		buff->r_ptr += (uint32_t)size_to_copy;
+		history_depth -= size_to_copy;
+
+		if (move_buffer) {
+			buff->r_ptr = buff->start_addr;
+			buff = buff->next;
+			move_buffer = 0;
+		}
+		if (size_to_copy)
+			comp_update_buffer_produce(sink, size_to_copy);
+	}
+
+	if (draining_done) {
```

Comment:
When the draining_done won't be set?
Why are we returning 0 for draining_done and 1 if not?

Reply:
With current implementation, in which we stay in while loop as long as `history_depth > 0 ` we can skip `draining_done `. It is a residue from earlier implementation which we could leave that loop in certain condition and get back at later time.
As for return value it has also be changed now to 0 right after we exit the function.